src/Nethermind/Nethermind.Network/P2P/RateLimitedPacketSender.cs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using DotNetty.Buffers;
+using DotNetty.Transport.Channels;
+using Nethermind.Logging;
+using Nethermind.Network.P2P.Messages;
+
+namespace Nethermind.Network.P2P;
+
+public class RateLimitedPacketSender : ChannelHandlerAdapter, IPacketSender, IDisposable
+{
+    private readonly int _byteLimit;
+    private readonly TimeSpan _throttleTime;
+    private readonly IMessageSerializationService _messageSerializationService;
+    private readonly ILogger _logger;
+    private readonly Channel<IByteBuffer> _channel;
+
+    private IChannelHandlerContext? _context;
+    private Task? _processor;
+
+    public RateLimitedPacketSender(int byteLimit, TimeSpan throttleTime, IMessageSerializationService messageSerializationService, ILogManager logManager)
+    {
+        _byteLimit = byteLimit;
+        _throttleTime = throttleTime;
+        _messageSerializationService = messageSerializationService;
+        _logger = logManager.GetClassLogger<RateLimitedPacketSender>();
+        _channel = Channel.CreateUnbounded<IByteBuffer>(new UnboundedChannelOptions
+        {
+            SingleReader = true,
+        });
+    }
+
+    public void Init()
+    {
+        _processor = Task.Run(async () =>
+        {
+            TimeSpan durationSoFar = TimeSpan.Zero;
+            int bytesSentSoFar = 0;
+
+            while (await _channel.Reader.WaitToReadAsync())
+            {
+                while (_channel.Reader.TryRead(out IByteBuffer buffer))
+                {
+                    try
+                    {
+                        int bytesToSend = buffer.ReadableBytes;
+                        if (bytesSentSoFar + bytesToSend > _byteLimit)
+                        {
+                            if (_throttleTime - durationSoFar > TimeSpan.Zero)
```

Comment:
if we achieve `durationSoFar > _throttleTime `then we will never rest the value of `durationSoFar` to 0, so we will never throttle again
Perhaps, I am missing something

Reply:
Added a test for this scenario (see https://github.com/NethermindEth/nethermind/pull/6292/files#diff-d20ccee1c0d0a38c4e7cfa501ccbc7c964119335f8bca3e7f291585355274f80R172) and fixed the potential issue.